head/src/main/kotlin/io/qalipsis/core/head/campaign/scheduler/DefaultCampaignSchedulerImpl.kt

```diff
@@ -63,9 +63,10 @@ internal class DefaultCampaignSchedulerImpl(
     }
 
     override suspend fun schedule(campaignKey: CampaignKey, instant: Instant) {
-        coroutineScope.launch {
+        val scheduleJob = coroutineScope.launch {
             schedulingExecution(campaignKey, instant)
         }
+        campaignPreparator.campaignKeyToJobMap[campaignKey] = scheduleJob
```

Comment:
Ideally, you should not access to a property that way. But why isn't the campaignKeyToJobMap in this class, as well as the cancellation? (you already added the update, cancel could be done here as well)

Reply:
The map was in the class as well earlier. It was part of the earlier implementation that resulted in the circular dependency issue

Reply:
https://github.com/qalipsis/qalipsis-oss/pull/179#discussion_r1280524447